{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Antigen Table Condensing\n",
    "This notebook is to fetch all relevant data for the antigens presented in supplementary table 7 by matching with the 6 prior antigen tables. This is meant to make finding information about antigens (such as their sequence, function, etc...) easier for screening."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's create a class to represent an antigen\n",
    "class Antigen:\n",
    "    def __init__(self):\n",
    "        # Identifier of the protein\n",
    "        self.protein_id = None\n",
    "        \n",
    "        # Start position of the antigen in the protein\n",
    "        self.start_position = None\n",
    "        \n",
    "        # Sequence of the antigen itself\n",
    "        self.antigen_sequence = None\n",
    "        \n",
    "        # Sequence of the protein that the antigen is a part of\n",
    "        self.protein_sequence = None \n",
    "        \n",
    "        # Smp protein isoforms\n",
    "        self.smp_isoform = None\n",
    "        \n",
    "        # Smp protein\n",
    "        self.smp_protein = None\n",
    "        \n",
    "        # Subcellular localization of the protein\n",
    "        self.subcellular_localization = None\n",
    "        \n",
    "        # Associated stage\n",
    "        self.stage = None\n",
    "        \n",
    "        # Enrichment information\n",
    "        self.enrichment = None\n",
    "        \n",
    "        # Smp id\n",
    "        self.smp_id = None\n",
    "        \n",
    "        # Function\n",
    "        self.function = None\n",
    "        \n",
    "        # Single cell localization\n",
    "        self.single_cell_localization = None\n",
    "        \n",
    "    # format antigens like they are presented in the tables\n",
    "    def __str__(self):\n",
    "        return f'{self.protein_id}|{self.start_position}'\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return f'{self.protein_id}|{self.start_position}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collecting Information from Phage Display Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# first, read in the supplementary table 7\n",
    "df7 = pd.read_excel('Tables/Suppl. Table 7 - Vaccine selected peptides primers and amplification conditions.xlsx', header=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Code</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>(Sequence (5'-&gt;3')</th>\n",
       "      <th>Product lenght (nt)</th>\n",
       "      <th>TM super-fi</th>\n",
       "      <th>Concentration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Smp_075800.1|766</td>\n",
       "      <td>Forward primer</td>\n",
       "      <td>CTGACCATGGCATCCAGCTGGATCAAC</td>\n",
       "      <td>208.0</td>\n",
       "      <td>72°C</td>\n",
       "      <td>0.5 uM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Reverse primer</td>\n",
       "      <td>TCATGCTGCAGACGATGGCGGCTAGCG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Smp_075800.1|613</td>\n",
       "      <td>Forward primer</td>\n",
       "      <td>CTGACCATGGCATCCAGAATCTTTCTA</td>\n",
       "      <td>208.0</td>\n",
       "      <td>67.9°C</td>\n",
       "      <td>0.5 uM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Reverse primer</td>\n",
       "      <td>TCATGCTGCAGACGATGCTCTTTATACTG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Smp_179950.1|307</td>\n",
       "      <td>Forward primer</td>\n",
       "      <td>CTGACCATGGCATCCAGAAATCGAT</td>\n",
       "      <td>208.0</td>\n",
       "      <td>70.2°C</td>\n",
       "      <td>0.5 uM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Code      Unnamed: 1             (Sequence (5'->3')  \\\n",
       "0   Smp_075800.1|766  Forward primer    CTGACCATGGCATCCAGCTGGATCAAC   \n",
       "1                NaN  Reverse primer    TCATGCTGCAGACGATGGCGGCTAGCG   \n",
       "2   Smp_075800.1|613  Forward primer    CTGACCATGGCATCCAGAATCTTTCTA   \n",
       "3                NaN  Reverse primer  TCATGCTGCAGACGATGCTCTTTATACTG   \n",
       "4   Smp_179950.1|307  Forward primer      CTGACCATGGCATCCAGAAATCGAT   \n",
       "\n",
       "   Product lenght (nt) TM super-fi Concentration  \n",
       "0                208.0       72°C         0.5 uM  \n",
       "1                  NaN         NaN           NaN  \n",
       "2                208.0      67.9°C        0.5 uM  \n",
       "3                  NaN         NaN           NaN  \n",
       "4                208.0      70.2°C        0.5 uM  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# look at the df\n",
    "df7[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Smp_075800.1|766', Smp_075800.1|766),\n",
       " ('Smp_075800.1|613', Smp_075800.1|613),\n",
       " ('Smp_179950.1|307', Smp_179950.1|307),\n",
       " ('Smp_158420.1|307', Smp_158420.1|307),\n",
       " ('MEG4.161011091|307', MEG4.161011091|307)]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fetch all of the relevant antigens from the first column\n",
    "antigens = {} # map of antigen name to antigen object\n",
    "for name in df7['Code'].dropna().unique():\n",
    "    antigen = Antigen()\n",
    "    antigen.protein_id, antigen.start_position = name.strip().split('|')\n",
    "    antigens[name.strip()] = antigen\n",
    "    \n",
    "# look at the antigens\n",
    "list(antigens.items())[:5] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Peptide.ID</th>\n",
       "      <th>Peptide.sequence</th>\n",
       "      <th>Rhwk08pi</th>\n",
       "      <th>Rhwk10pi</th>\n",
       "      <th>Rhwk12pi</th>\n",
       "      <th>Rhwk1pc</th>\n",
       "      <th>Rhwk4pc</th>\n",
       "      <th>Hmwk12pi</th>\n",
       "      <th>Hmwk22pi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c10378-g1-i10123|1</td>\n",
       "      <td>MNLPILGRHSTKLFVFFNCLFITFGIIVIAFGMKLPTKLFRYKRIL...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c10378-g1-i10123|154</td>\n",
       "      <td>PIFPTATFSGSLSILATCIGILGILNERNMIMYLHLCTLTIVMITE...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>c10378-g1-i10123|307</td>\n",
       "      <td>RNMIMYLHLCTLTIVMITEIGIASSSKLMKDQFFTEAHRSLNMSVQ...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c10577-g1-i10123|1</td>\n",
       "      <td>TGFKVGINYQPPTVVPGGDLAKVQRAVCMLSNTTAIAEAWARLDHK...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c10577-g1-i10123|154</td>\n",
       "      <td>WARLDHKFDLMYAKRAFVHWYVGEGMEEGEFSEAREDMAALEKDYE...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Peptide.ID                                   Peptide.sequence  \\\n",
       "0    c10378-g1-i10123|1  MNLPILGRHSTKLFVFFNCLFITFGIIVIAFGMKLPTKLFRYKRIL...   \n",
       "1  c10378-g1-i10123|154  PIFPTATFSGSLSILATCIGILGILNERNMIMYLHLCTLTIVMITE...   \n",
       "2  c10378-g1-i10123|307  RNMIMYLHLCTLTIVMITEIGIASSSKLMKDQFFTEAHRSLNMSVQ...   \n",
       "3    c10577-g1-i10123|1  TGFKVGINYQPPTVVPGGDLAKVQRAVCMLSNTTAIAEAWARLDHK...   \n",
       "4  c10577-g1-i10123|154  WARLDHKFDLMYAKRAFVHWYVGEGMEEGEFSEAREDMAALEKDYE...   \n",
       "\n",
       "  Rhwk08pi Rhwk10pi Rhwk12pi Rhwk1pc Rhwk4pc Hmwk12pi Hmwk22pi  \n",
       "0       -        -        -       -       -        -        -   \n",
       "1       -        -        -       -       -        -        -   \n",
       "2       -        -        -       -       -        -        -   \n",
       "3       -        -        -       -       -        -        -   \n",
       "4       -        -        -       -       -        -        -   "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now, get the antigen sequences from supplementary table 2\n",
    "df2 = pd.read_excel('Tables/Suppl. Table 2 - enriched peptides sequences.xlsx', header=1)\n",
    "df2[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Peptide.ID</th>\n",
       "      <th>Peptide.sequence</th>\n",
       "      <th>Rhwk08pi</th>\n",
       "      <th>Rhwk10pi</th>\n",
       "      <th>Rhwk12pi</th>\n",
       "      <th>Rhwk1pc</th>\n",
       "      <th>Rhwk4pc</th>\n",
       "      <th>Hmwk12pi</th>\n",
       "      <th>Hmwk22pi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1699</th>\n",
       "      <td>MEG151001110|154</td>\n",
       "      <td>VPAKTTRKHHNTLKTTLHHNTLKTTLCLTTTQYQQKSQHPNTTPSH...</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1700</th>\n",
       "      <td>MEG151001110|307</td>\n",
       "      <td>PAKTTRKHHNTLKTTLHHNTLKTTLCLTTTQYQQKSQHPNTTPSHT...</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>MEG15400100|154</td>\n",
       "      <td>HYVGKLSHHNTVPAKTTRKSQHTTATARHHNTLKTTLSHHNTVPAK...</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1968</th>\n",
       "      <td>MEG15410110|307</td>\n",
       "      <td>TTRKSQHTTATARHHNTLKTTLSHHNTVPAKTTRNHTDKTILIKPY...</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3255</th>\n",
       "      <td>MEG3.1710010|154</td>\n",
       "      <td>SNTGPQTFYSSTQDDFDKCRNDCGRVTSHHELCGDVCGGNHGGSFP...</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Peptide.ID                                   Peptide.sequence  \\\n",
       "1699  MEG151001110|154  VPAKTTRKHHNTLKTTLHHNTLKTTLCLTTTQYQQKSQHPNTTPSH...   \n",
       "1700  MEG151001110|307  PAKTTRKHHNTLKTTLHHNTLKTTLCLTTTQYQQKSQHPNTTPSHT...   \n",
       "1953   MEG15400100|154  HYVGKLSHHNTVPAKTTRKSQHTTATARHHNTLKTTLSHHNTVPAK...   \n",
       "1968   MEG15410110|307  TTRKSQHTTATARHHNTLKTTLSHHNTVPAKTTRNHTDKTILIKPY...   \n",
       "3255  MEG3.1710010|154  SNTGPQTFYSSTQDDFDKCRNDCGRVTSHHELCGDVCGGNHGGSFP...   \n",
       "\n",
       "     Rhwk08pi Rhwk10pi Rhwk12pi Rhwk1pc Rhwk4pc Hmwk12pi Hmwk22pi  \n",
       "1699       -       yes      yes     yes     yes       -        -   \n",
       "1700       -       yes      yes     yes     yes       -        -   \n",
       "1953       -       yes      yes     yes     yes       -        -   \n",
       "1968       -       yes      yes     yes     yes       -        -   \n",
       "3255       -       yes      yes     yes     yes      yes      yes  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract only rows that match antigens of interest\n",
    "df2 = df2[df2[\"Peptide.ID\"].isin(antigens.keys())]\n",
    "df2[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Smp_075800.1|766 LDQQYKEVKRETNLSHVQRYGDTRMGKLHVSEFQGSRDKSSTENDEPPMKPRHSIASR {'Rhwk08pi': True, 'Rhwk10pi': False, 'Rhwk12pi': False, 'Rhwk1pc': True, 'Rhwk4pc': True, 'Hmwk12pi': False, 'Hmwk22pi': False}\n",
      "Smp_075800.1|613 NLSIYATTAASPTESSYGTFCDDPTITTCLADLYSYDWIVDSQTHHLTQRTLDQQYKE {'Rhwk08pi': True, 'Rhwk10pi': False, 'Rhwk12pi': False, 'Rhwk1pc': False, 'Rhwk4pc': False, 'Hmwk12pi': False, 'Hmwk22pi': False}\n",
      "Smp_179950.1|307 KSIATIRDQSRCGSCCAFGAVEAMSERSCIQSGGKQNVELSAVDLEGIVTGSSKENNT {'Rhwk08pi': True, 'Rhwk10pi': False, 'Rhwk12pi': False, 'Rhwk1pc': False, 'Rhwk4pc': False, 'Hmwk12pi': False, 'Hmwk22pi': False}\n",
      "Smp_158420.1|307 AMSDRSCIQSGGKQNVELSAVDLLSCCEHCGDGFEGGFPALAWDYWVKEGIVTGSSKE {'Rhwk08pi': True, 'Rhwk10pi': False, 'Rhwk12pi': False, 'Rhwk1pc': False, 'Rhwk4pc': True, 'Hmwk12pi': False, 'Hmwk22pi': False}\n",
      "MEG4.161011091|307 ENKKDKHPTQKEQLPNLQHQRQINDGTSDKTSDTHTIKRQLPNLQHQRTTPKPTTPKT {'Rhwk08pi': True, 'Rhwk10pi': True, 'Rhwk12pi': True, 'Rhwk1pc': True, 'Rhwk4pc': True, 'Hmwk12pi': False, 'Hmwk22pi': False}\n"
     ]
    }
   ],
   "source": [
    "# now, fill in the antigen sequences\n",
    "for index, row in df2.iterrows():\n",
    "    antigen = antigens[row[\"Peptide.ID\"]]\n",
    "    \n",
    "    antigen.antigen_sequence = row[\"Peptide.sequence\"]\n",
    "    enrichment = {}\n",
    "    for col in (\"Rhwk08pi\", \"Rhwk10pi\", \"Rhwk12pi\", \"Rhwk1pc\", \"Rhwk4pc\", \"Hmwk12pi\", \"Hmwk22pi\"):\n",
    "        if row[col] == \"yes\":\n",
    "            enrichment[col] = True\n",
    "        else:\n",
    "            enrichment[col] = False\n",
    "    antigen.enrichment = enrichment\n",
    "            \n",
    "# peek at the antigens\n",
    "for antigen in list(antigens.values())[:5]:\n",
    "    print(antigen, antigen.antigen_sequence, antigen.enrichment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Smp_127350.1', 'MDGLSTSWTKSSVSDTCQRSNVNIHTMLRNPTYIDRVERR'),\n",
       " ('MEG3.1500100',\n",
       "  'MLFFALILIISLHSFDCVFTARETRDAALSCKKHCEGNNEYVTRYCGGLCSSNTGPQTFYCYLGCSHNASTQDDFDKCRNDCGRVNHGGSFPLCLYNCDQEHPREYERGYDECKCNIS'),\n",
       " ('Smp_172500.1',\n",
       "  'MASLHIHQDHVILVDYRIKYYCMIHPRQLIIIIVLVFICRIHHPQSTKLITTPNMNLTLIQQNRQIRNNSPPPLPPPPKHLIQMLGGDSSIYSTPHGSPQQVNTHKLMILNNSNSNTPNNNNISVDYSPNLLHVSSISKHESSTTGSFRQTDSGVTNMGGSLIDSEQYEPNTIVT'),\n",
       " ('MEG8.21201100',\n",
       "  'MFTIILIYVLYFIANAKFEHTTSGIRNPSKLSDSNASKTLSLKNLTDHYIHTPQKSNNGTSCNGKDTCKLPNPSQKGFTNTTSLPHTQSHNSTVAPSVPKPTRQEIPRSGTIVNGTKPTPGKPVVNGTKPTPGKPESSSIQ'),\n",
       " ('Smp_129820.1',\n",
       "  'MVEYKRRACQPRSQMLFAFTCLIIVCMLPRLYGLLYAGRGWYNQSTFYWLHAATISVYICLAFGELLIQFFSPFNEQNSLLDDKDACPELWASVPALWTFNWLTSTIWKGFRGQINSPSDMFHIRPEDSVNKSYGRFKRAWNRSYNLWFKRKEPETPFSNVMDDDVPLLMDFKENNEDTENVSNDANMLNATPDVDILSPSTTNNNSNSYNLEVCKQSTIKASPARATWGLFYALISSFGLRLFCGWIFISAAVFFNYSSPLLLGMLLRFLSKPEAPLWQGYFIAFSMLFLQSISVLVDQKGFYSCLSLGISVRSALTSAIYRKSLRLSSEARSQYTTGELINLLSVDVNRIKELFMFSFLVWDAFIEFGLSFILLWRQLGSAAIAGVGFLLLVLPLNCLLIWATQKFEIREMKYKDKRMKCLGEVLAAIRVVKLYAWEKAFQSQVKSIRQSELTELLRVALGWGLCHVVWNLAPYIILLITFVTYLREFLFANHNVLLDNWVGNTTDIPTPQLLTPERIFVSVSLFNLLRSPLLLLPWSLSSSIMAFVSIRRLGLFLLADELEHYSSDKIYPLDENAIIFENASFSWTKTGPLALQKDSFCPTYRSLTLMPNDNDFQPTINRSDIYLLDDPLSAVDTHVGKYLFDHVLGPNGLLKDKTRIMTTNSFHWLSLADWIIVLNTNGEITQSGTYNEVVNNNSGHFTNYLESIEKNKINDETSNQRKISFTTPFNRSSSVVVVMATDGALNDLIDFRSEQEVTHRRSAASISTTDLNASTSDNDKLENDDVECGKFMTDEEIMHGHVSWSTYWEYFRARSVSVTFISVLSYAGFLAFQNTTIISNETARQILIEQIKDRILYYIIGYAWAGLGQTVLILLFALLNAYSNLRASNLLHQHLLSKILHAPASFFDHTPLGRILNRFSNDVDNLDHTIPNSFSDTIGTTGDVLISLMIVTVTLRPFGIGLAVIIPVFIFCITVLVFYMPSVRQTRRLDANTRSPLLTNYSETSASSLGVTVVRAFKRDKEFIAKSDKLIDANAVFEYVRFVANRWLDVHLNLSCSFMVFVCAIILVAFRSKISPGIAGLIIVYALQVFDSLTWVIKQLSQLETSSVSLERICEYIRVEQEADWDHGVDSPPPIDWPRPCCEITFNKATVQYHLPTKNKVTDNHDLSVSQKTSSNSGRLVTALKSIDLKLSGNPQERRIGIVGRTGAGKSSLASSIFRLIEPTILEEDRLDTHRTSTKRGPIVVDGVDLSRIGLHELRSRFSILPQEPIIFAGTLRFNLDPFGKLPDIDLWRAIESAHLADWVRSTNAGLDYECGEGGANLSAGQRQLVCLARVFLSSGGRVRLLILDEATAAMDPSTDNLVMNTVIGDIFKDATVIIIAHRLSTVMNTDRIVVLDHGQVIETGHPQDLLNNPNSHFSIMNKVKQS')]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now, get the protein sequences from the customReferenceProteinsDatabase.fa file\n",
    "protein_sequences = {}\n",
    "with open('customReferenceProteinsDatabase.fa') as f:\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            protein_id = line[1:].strip()\n",
    "            protein_sequences[protein_id] = ''\n",
    "        else:\n",
    "            protein_sequences[protein_id] += line.strip()\n",
    "\n",
    "list(protein_sequences.items())[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Smp_075800.1|766 MMLFSLFLISILHILLVKCQLDTNYEVSDETVSDNNKWAVLVAGSNGYFNYRHQADVCHAYHVLRSKGIKPEHIITMMYDDIAYNLMNPFPGKLFNDYNHKDWYKGVVIDYRGKKVNSKTFLKVLKGDKSAGGKVLKSGKNDDVFIYFTDHGAPGLIAFPDDELYAKEFMSTLKYLHSHKRYSKLVIYIEACESGSMFQRILPSNLSIYATTAASPTESSYGTFCDDPTITTCLADLYSYDWIVDSQTHHLTQRTLDQQYKEVKRETNLSHVQRYGDTRMGKLHVSEFQGSRDKSSTENDEPPMKPRHSIASRDIPLHTLHRQIMMTNNAEDKSFLMQILGLKLKRRDLIEDTMKLIVKVMNNEEIPNTKATIDQTLDCTESVYEQFKSKCFTLQQAPEVGGHFSTLYNYCADGYTAETINEAIIKICG\n",
      "Smp_075800.1|613 MMLFSLFLISILHILLVKCQLDTNYEVSDETVSDNNKWAVLVAGSNGYFNYRHQADVCHAYHVLRSKGIKPEHIITMMYDDIAYNLMNPFPGKLFNDYNHKDWYKGVVIDYRGKKVNSKTFLKVLKGDKSAGGKVLKSGKNDDVFIYFTDHGAPGLIAFPDDELYAKEFMSTLKYLHSHKRYSKLVIYIEACESGSMFQRILPSNLSIYATTAASPTESSYGTFCDDPTITTCLADLYSYDWIVDSQTHHLTQRTLDQQYKEVKRETNLSHVQRYGDTRMGKLHVSEFQGSRDKSSTENDEPPMKPRHSIASRDIPLHTLHRQIMMTNNAEDKSFLMQILGLKLKRRDLIEDTMKLIVKVMNNEEIPNTKATIDQTLDCTESVYEQFKSKCFTLQQAPEVGGHFSTLYNYCADGYTAETINEAIIKICG\n",
      "Smp_179950.1|307 MLISVLYIASLISHLEAHISIKNEKFEPLSDDIISYINEHPNAGWRAEKSNRFHSLDDARFQLGARREEPDLRRTRRPTVDHNDWNVEIPSSFDSRKKWPRCKSIATIRDQSRCGSCCAFGAVEAMSERSCIQSGGKQNVELSAVDLEGIVTGSSKENNTGCEPYPFPKCEHFTKGQYPPCGSKIYKTPRCKTTCQKRYKTSYAQDKHRAIQKEIMKYGPVEASFTVYEDFLNYKSGIYKHITGETLGGHAIRIIGWGVENKTPYWLIANSWNEDWGENGYFRIVRGRDECSIESEVTAGRIN\n",
      "Smp_158420.1|307 MLISVLCIASLITHLEAHISIKNEKFEPLSHDIISYINKHLDARREESDLRRKRRPIVDHNDWNVEIPSNFDSRKKWPGCKSIATIRDQSRCGSSWAFGAVEAMSDRSCIQSGGKQNVELSAVDLLSCCEHCGDGFEGGFPALAWDYWVKEGIVTGSSKENHTSCQPYPFPKCEHHTKGKYPACFEEIYKTPNCENTCQKSYKTPYAQDKHRGKSRYNVKNDEKAIQKEIMKYGPVEANFIVYEDFLNYKSGIYKHITGKLVSWHAIRIIGWGVENNTPYWLIPNSWNEDWGENGNFRILRGRHECSIESEVTAGRINE\n",
      "MEG4.161011091|307 MNIYLIGILCIVGLIISQGSTANGSPLDDRFNDVNTINKKQFTEEEFSRLINSMLKEYIEDKNVDIRIIENKKDKHPTQKEQLPNLQHQRQINDGTSDKTSDTHTIKRQLPNLQHQRTTPKPTTPKT\n"
     ]
    }
   ],
   "source": [
    "# update the antigens with the protein sequences\n",
    "for antigen in antigens.values():\n",
    "    protein_id = antigen.protein_id\n",
    "    antigen.protein_sequence = protein_sequences[protein_id]\n",
    "    \n",
    "# peek at the antigens\n",
    "for antigen in list(antigens.values())[:5]:\n",
    "    print(antigen, antigen.protein_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# double check that the antigen sequence is always in the protein sequence\n",
    "for antigen in antigens.values():\n",
    "    if not antigen.antigen_sequence in antigen.protein_sequence:\n",
    "        print(\"WARNING: Antigen sequence not found in protein sequence. An error may have occurred.\")\n",
    "        print(antigen, antigen.antigen_sequence, antigen.protein_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Enriched_peptide ID</th>\n",
       "      <th>Smp_ID (genome V7)</th>\n",
       "      <th>Wk08pi</th>\n",
       "      <th>Wk10pi</th>\n",
       "      <th>Wk12pi</th>\n",
       "      <th>Wk1pc</th>\n",
       "      <th>Wk4pc</th>\n",
       "      <th>Hamster wk12+wk22</th>\n",
       "      <th>Function</th>\n",
       "      <th>Single-cell localization</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Smp_067060.1|460</td>\n",
       "      <td>Smp_067060</td>\n",
       "      <td>yes</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>yes</td>\n",
       "      <td>Cathepsin B-like peptidase (C01 family)</td>\n",
       "      <td>Gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Smp_179950.1|307</td>\n",
       "      <td>Smp_179950</td>\n",
       "      <td>yes</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>Cathepsin B-like peptidase (C01 family)</td>\n",
       "      <td>Gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Smp_075800.1|766</td>\n",
       "      <td>Smp_075800</td>\n",
       "      <td>yes</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>--</td>\n",
       "      <td>Asparaginyl endopeptidase</td>\n",
       "      <td>Gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Smp_075800.1|613</td>\n",
       "      <td>Smp_075800</td>\n",
       "      <td>yes</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>Asparaginyl endopeptidase</td>\n",
       "      <td>Gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Smp_158420.1|307</td>\n",
       "      <td>Smp_344940</td>\n",
       "      <td>yes</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>yes</td>\n",
       "      <td>--</td>\n",
       "      <td>cathepsin B peptidase</td>\n",
       "      <td>Gut</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Enriched_peptide ID Smp_ID (genome V7) Wk08pi Wk10pi Wk12pi Wk1pc Wk4pc  \\\n",
       "0    Smp_067060.1|460         Smp_067060    yes     --     --    --    --   \n",
       "1    Smp_179950.1|307         Smp_179950    yes     --     --    --    --   \n",
       "2    Smp_075800.1|766         Smp_075800    yes     --     --   yes   yes   \n",
       "3    Smp_075800.1|613         Smp_075800    yes     --     --    --    --   \n",
       "4    Smp_158420.1|307         Smp_344940    yes     --     --    --   yes   \n",
       "\n",
       "  Hamster wk12+wk22                                 Function  \\\n",
       "0               yes  Cathepsin B-like peptidase (C01 family)   \n",
       "1                --  Cathepsin B-like peptidase (C01 family)   \n",
       "2                --                Asparaginyl endopeptidase   \n",
       "3                --                Asparaginyl endopeptidase   \n",
       "4                --                    cathepsin B peptidase   \n",
       "\n",
       "  Single-cell localization  \n",
       "0                      Gut  \n",
       "1                      Gut  \n",
       "2                      Gut  \n",
       "3                      Gut  \n",
       "4                      Gut  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now, we can get more information from the remaining supplementary tables\n",
    "df3 = pd.read_excel('Tables/Suppl. Table 3 - Extracellular proteins enrichment, function and localization.xlsx', header=1)\n",
    "df3[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Smp_075800.1|766 Smp_075800 Asparaginyl endopeptidase Gut\n",
      "Smp_075800.1|613 Smp_075800 Asparaginyl endopeptidase Gut\n",
      "Smp_179950.1|307 Smp_179950 Cathepsin B-like peptidase (C01 family) Gut\n",
      "Smp_158420.1|307 Smp_344940 cathepsin B peptidase Gut\n",
      "MEG4.161011091|307 Smp_307220 MEG-4 (10.3) family Oesophageal gland\n"
     ]
    }
   ],
   "source": [
    "# update the antigens with the function and localization\n",
    "\n",
    "for index, row in df3.iterrows():\n",
    "    id = row[\"Enriched_peptide ID\"].strip()\n",
    "    antigen_ids = []\n",
    "    \n",
    "    # check if this is a protein or antigen\n",
    "    if \"|\" in id: # antigen\n",
    "        antigen_ids = [id]\n",
    "    else: # protein\n",
    "        protein_id = id.split(\" \")[0].replace(\"-\",\"\")\n",
    "        antigen_ids = [str(antigen) for antigen in antigens.values() if protein_id in antigen.protein_id]\n",
    "    \n",
    "    # update the antigens\n",
    "    for antigen_id in antigen_ids:\n",
    "        # skip any irrelevant antigens\n",
    "        if antigen_id not in antigens:\n",
    "            continue\n",
    "        \n",
    "        antigen = antigens[antigen_id]\n",
    "        antigen.smp_id = row[\"Smp_ID (genome V7)\"]\n",
    "        antigen.function = row[\"Function\"]\n",
    "        antigen.single_cell_localization = row[\"Single-cell localization\"]\n",
    "        \n",
    "# peek at the antigens\n",
    "for antigen in list(antigens.values())[:5]:\n",
    "    print(antigen, antigen.smp_id, antigen.function, antigen.single_cell_localization)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Peptide ID</th>\n",
       "      <th>Smp protein \\nIsoforms (v7)</th>\n",
       "      <th>Smp \\nProtein (v7)</th>\n",
       "      <th>Subcellular Localization</th>\n",
       "      <th>stage</th>\n",
       "      <th>Hmwk12</th>\n",
       "      <th>Hmwk22</th>\n",
       "      <th>Rhwk08</th>\n",
       "      <th>Rhwk10</th>\n",
       "      <th>Rhwk12</th>\n",
       "      <th>Rhwk43</th>\n",
       "      <th>Rhwk46</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c11762-g1-i10123|154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c12986-g1-i10123|307</td>\n",
       "      <td>Smp_306120.1</td>\n",
       "      <td>Smp_306120</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>juvenile</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>c1504-g1-i10123|613</td>\n",
       "      <td>Smp_128360.1</td>\n",
       "      <td>Smp_128360</td>\n",
       "      <td>Mitochondrion</td>\n",
       "      <td>adult females</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c18831-g1-i10123|307</td>\n",
       "      <td>Smp_178360.1</td>\n",
       "      <td>Smp_178360</td>\n",
       "      <td>Cytoplasm|Nucleus</td>\n",
       "      <td>juvenile</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c19029-g1-i10123|613</td>\n",
       "      <td>Smp_302920.2</td>\n",
       "      <td>Smp_302920</td>\n",
       "      <td>Cytoplasm</td>\n",
       "      <td>juvenile</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Peptide ID Smp protein \\nIsoforms (v7) Smp \\nProtein (v7)  \\\n",
       "0  c11762-g1-i10123|154                         NaN                NaN   \n",
       "1  c12986-g1-i10123|307                Smp_306120.1         Smp_306120   \n",
       "2   c1504-g1-i10123|613                Smp_128360.1         Smp_128360   \n",
       "3  c18831-g1-i10123|307                Smp_178360.1         Smp_178360   \n",
       "4  c19029-g1-i10123|613                Smp_302920.2         Smp_302920   \n",
       "\n",
       "  Subcellular Localization          stage Hmwk12 Hmwk22 Rhwk08 Rhwk10 Rhwk12  \\\n",
       "0                      NaN            NaN     -     yes     -      -      -    \n",
       "1                Cytoplasm       juvenile     -     yes     -      -      -    \n",
       "2            Mitochondrion  adult females     -     yes     -      -      -    \n",
       "3        Cytoplasm|Nucleus       juvenile     -     yes     -      -      -    \n",
       "4                Cytoplasm       juvenile     -     yes     -     yes    yes   \n",
       "\n",
       "  Rhwk43 Rhwk46  \n",
       "0     -      -   \n",
       "1     -      -   \n",
       "2     -      -   \n",
       "3     -      -   \n",
       "4    yes    yes  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now for the subcellular localization and more information\n",
    "df5 = pd.read_excel('Tables/Suppl. Table 5 - Enriched peptides in each analysis and their localization and stage expression.xls', header=2)\n",
    "df5[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Peptide ID</th>\n",
       "      <th>Smp protein \\nIsoforms (v7)</th>\n",
       "      <th>Smp \\nProtein (v7)</th>\n",
       "      <th>Subcellular Localization</th>\n",
       "      <th>stage</th>\n",
       "      <th>Hmwk12</th>\n",
       "      <th>Hmwk22</th>\n",
       "      <th>Rhwk08</th>\n",
       "      <th>Rhwk10</th>\n",
       "      <th>Rhwk12</th>\n",
       "      <th>Rhwk43</th>\n",
       "      <th>Rhwk46</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>MEG151001110|154</td>\n",
       "      <td>Smp_010550.2</td>\n",
       "      <td>Smp_010550</td>\n",
       "      <td>Extracellular</td>\n",
       "      <td>adult females</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>MEG151001110|307</td>\n",
       "      <td>Smp_010550.2</td>\n",
       "      <td>Smp_010550</td>\n",
       "      <td>Extracellular</td>\n",
       "      <td>adult females</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>MEG15400100|154</td>\n",
       "      <td>Smp_010550.2</td>\n",
       "      <td>Smp_010550</td>\n",
       "      <td>Extracellular</td>\n",
       "      <td>adult females</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>MEG15410110|307</td>\n",
       "      <td>Smp_010550.2</td>\n",
       "      <td>Smp_010550</td>\n",
       "      <td>Extracellular</td>\n",
       "      <td>adult females</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>MEG3.1710010|154</td>\n",
       "      <td>Smp_138080.1</td>\n",
       "      <td>Smp_138080</td>\n",
       "      <td>Extracellular</td>\n",
       "      <td>somula</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>-</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Peptide ID Smp protein \\nIsoforms (v7) Smp \\nProtein (v7)  \\\n",
       "45   MEG151001110|154                Smp_010550.2         Smp_010550   \n",
       "46   MEG151001110|307                Smp_010550.2         Smp_010550   \n",
       "78    MEG15400100|154                Smp_010550.2         Smp_010550   \n",
       "80    MEG15410110|307                Smp_010550.2         Smp_010550   \n",
       "137  MEG3.1710010|154                Smp_138080.1         Smp_138080   \n",
       "\n",
       "    Subcellular Localization          stage Hmwk12 Hmwk22 Rhwk08 Rhwk10  \\\n",
       "45             Extracellular  adult females     -      -      -     yes   \n",
       "46             Extracellular  adult females     -      -      -     yes   \n",
       "78             Extracellular  adult females     -      -      -     yes   \n",
       "80             Extracellular  adult females     -      -      -     yes   \n",
       "137            Extracellular         somula    yes    yes     -     yes   \n",
       "\n",
       "    Rhwk12 Rhwk43 Rhwk46  \n",
       "45     yes    yes    yes  \n",
       "46     yes    yes    yes  \n",
       "78     yes    yes    yes  \n",
       "80     yes    yes    yes  \n",
       "137    yes    yes    yes  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# just keep the relevant rows\n",
    "df5 = df5[df5[\"Peptide ID\"].isin(antigens.keys())]\n",
    "df5[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Smp_075800.1|766 Smp_075800.1 Smp_075800 Extracellular adult females {'Rhwk08pi': True, 'Rhwk10pi': False, 'Rhwk12pi': False, 'Rhwk1pc': True, 'Rhwk4pc': True, 'Hmwk12pi': False, 'Hmwk22pi': False, 'Hmwk12': False, 'Hmwk22': False, 'Rhwk08': True, 'Rhwk10': False, 'Rhwk12': False, 'Rhwk43': True, 'Rhwk46': True}\n",
      "Smp_075800.1|613 Smp_075800.1 Smp_075800 Extracellular adult females {'Rhwk08pi': True, 'Rhwk10pi': False, 'Rhwk12pi': False, 'Rhwk1pc': False, 'Rhwk4pc': False, 'Hmwk12pi': False, 'Hmwk22pi': False, 'Hmwk12': False, 'Hmwk22': False, 'Rhwk08': True, 'Rhwk10': False, 'Rhwk12': False, 'Rhwk43': False, 'Rhwk46': False}\n",
      "Smp_179950.1|307 Smp_067060.1 Smp_067060 Extracellular adult females {'Rhwk08pi': True, 'Rhwk10pi': False, 'Rhwk12pi': False, 'Rhwk1pc': False, 'Rhwk4pc': False, 'Hmwk12pi': False, 'Hmwk22pi': False, 'Hmwk12': False, 'Hmwk22': False, 'Rhwk08': True, 'Rhwk10': False, 'Rhwk12': False, 'Rhwk43': False, 'Rhwk46': False}\n",
      "Smp_158420.1|307 Smp_344940.1 Smp_344940 Extracellular|Lysosome/Vacuole somula {'Rhwk08pi': True, 'Rhwk10pi': False, 'Rhwk12pi': False, 'Rhwk1pc': False, 'Rhwk4pc': True, 'Hmwk12pi': False, 'Hmwk22pi': False, 'Hmwk12': False, 'Hmwk22': False, 'Rhwk08': True, 'Rhwk10': False, 'Rhwk12': False, 'Rhwk43': False, 'Rhwk46': True}\n",
      "MEG4.161011091|307 Smp_307220.3 Smp_307220 Extracellular juvenile {'Rhwk08pi': True, 'Rhwk10pi': True, 'Rhwk12pi': True, 'Rhwk1pc': True, 'Rhwk4pc': True, 'Hmwk12pi': False, 'Hmwk22pi': False, 'Hmwk12': False, 'Hmwk22': False, 'Rhwk08': True, 'Rhwk10': True, 'Rhwk12': True, 'Rhwk43': True, 'Rhwk46': True}\n"
     ]
    }
   ],
   "source": [
    "# update the antigens\n",
    "for index, row in df5.iterrows():\n",
    "    antigen = antigens[row[\"Peptide ID\"]]\n",
    "    antigen.smp_isoform = row[\"Smp protein \\nIsoforms (v7)\"]\n",
    "    antigen.smp_protein = row[\"Smp \\nProtein (v7)\"]\n",
    "    antigen.subcellular_localization = row[\"Subcellular Localization\"]\n",
    "    antigen.stage = row[\"stage\"]\n",
    "    \n",
    "    for col in (\"Hmwk12\", \"Hmwk22\", \"Rhwk08\", \"Rhwk10\", \"Rhwk12\", \"Rhwk43\", \"Rhwk46\"):\n",
    "        if row[col] == \"yes\":\n",
    "            antigen.enrichment[col] = True\n",
    "        else:\n",
    "            antigen.enrichment[col] = False\n",
    "            \n",
    "# peek at the antigens\n",
    "for antigen in list(antigens.values())[:5]:\n",
    "    print(antigen, antigen.smp_isoform, antigen.smp_protein, antigen.subcellular_localization, antigen.stage, antigen.enrichment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now, let's write this all out to a new table\n",
    "with open('antigens.csv', 'w') as f:\n",
    "    f.write('Protein ID,Start Position,Antigen Sequence,Protein Sequence,Smp Isoform,Smp Protein,Subcellular Localization,Stage,Function,Single Cell Localization\\n')\n",
    "    for antigen in antigens.values():\n",
    "        f.write(f'{antigen.protein_id},{antigen.start_position},{antigen.antigen_sequence},{antigen.protein_sequence},{antigen.smp_isoform},{antigen.smp_protein},{antigen.subcellular_localization},{antigen.stage},{antigen.function},{antigen.single_cell_localization}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BLASTing Antigens and Proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /opt/anaconda3/lib/python3.9/site-packages (1.83)\n",
      "Requirement already satisfied: numpy in /opt/anaconda3/lib/python3.9/site-packages (from biopython) (1.21.3)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install biopython\n",
    "from Bio.Blast import NCBIWWW\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Antigen Smp_075800.1|766 already BLASTed\n",
      "Antigen Smp_075800.1|613 already BLASTed\n",
      "Antigen Smp_179950.1|307 already BLASTed\n",
      "Antigen Smp_158420.1|307 already BLASTed\n",
      "Antigen MEG4.161011091|307 already BLASTed\n",
      "Antigen MEG4.1610110|154 already BLASTed\n",
      "Antigen MEG4.11101100|154 already BLASTed\n",
      "Antigen MEG4.1700100|154 already BLASTed\n",
      "Antigen MEG4.1200001|307 already BLASTed\n",
      "Antigen MEG4.11711010172|154 already BLASTed\n",
      "Antigen MEG4.1700110|154 already BLASTed\n",
      "Antigen MEG4.11601101172|154 already BLASTed\n",
      "Antigen MEG4.11710010172|154 already BLASTed\n",
      "Antigen MEG4.1901110|307 already BLASTed\n",
      "Antigen MEG4.1610100|154 already BLASTed\n",
      "Antigen MEG4.1711010|154 already BLASTed\n",
      "Antigen MEG4.1701100|307 already BLASTed\n",
      "Antigen MEG4.11500110|154 already BLASTed\n",
      "Antigen MEG8.2400100|154 already BLASTed\n",
      "Antigen MEG8.21500100|154 already BLASTed\n",
      "Antigen MEG8.21501010|154 already BLASTed\n",
      "Antigen MEG8.21510010181|154 already BLASTed\n",
      "Antigen MEG8.21401000|154 already BLASTed\n",
      "Antigen MEG8.21200110|154 already BLASTed\n",
      "Antigen MEG8.21411000|154 already BLASTed\n",
      "Antigen MEG8.21510100|154 already BLASTed\n",
      "Antigen MEG8.21200100|154 already BLASTed\n",
      "Antigen MEG8.2511000|154 already BLASTed\n",
      "Antigen MEG8.2901100|154 already BLASTed\n",
      "Antigen MEG8.2511010|154 already BLASTed\n",
      "Antigen MEG3.1711000|154 already BLASTed\n",
      "Antigen MEG3.1710010|154 already BLASTed\n",
      "Antigen MEG15410110|307 already BLASTed\n",
      "Antigen MEG151001110|154 already BLASTed\n",
      "Antigen MEG151001110|307 already BLASTed\n",
      "Antigen MEG15400100|154 already BLASTed\n"
     ]
    }
   ],
   "source": [
    "# Ensure the output directory exists\n",
    "os.makedirs('blast_results/antigens', exist_ok=True)\n",
    "\n",
    "# let's blast the antigen sequences\n",
    "def blast_antigen(antigen):\n",
    "    print(f'BLASTing antigen {antigen}')\n",
    "    result_handle = NCBIWWW.qblast(\"blastp\", \"nr\", antigen.antigen_sequence)\n",
    "    with open(f'blast_results/antigens/{antigen}.xml', 'w') as f:\n",
    "        f.write(result_handle.read())\n",
    "    result_handle.close()\n",
    "    print(f'BLASTing antigen {antigen} done')\n",
    "    \n",
    "\n",
    "for antigen in antigens.values():\n",
    "    # check if this antigen has been blasted already\n",
    "    if os.path.exists(f'blast_results/antigens/{antigen}.xml'):\n",
    "        print(f'Antigen {antigen} already BLASTed')\n",
    "        continue\n",
    "    \n",
    "    blast_antigen(antigen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein MEG8.21200100 already BLASTed\n",
      "Protein MEG4.11601101172 already BLASTed\n",
      "Protein MEG4.1711010 already BLASTed\n",
      "Protein MEG4.11710010172 already BLASTed\n",
      "Protein MEG8.2400100 already BLASTed\n",
      "Protein MEG8.2901100 already BLASTed\n",
      "Protein Smp_179950.1 already BLASTed\n",
      "Protein MEG4.1901110 already BLASTed\n",
      "Protein MEG4.11500110 already BLASTed\n",
      "Protein MEG8.21510100 already BLASTed\n",
      "Protein MEG8.21510010181 already BLASTed\n",
      "Protein MEG8.21401000 already BLASTed\n",
      "Protein MEG4.1610100 already BLASTed\n",
      "Protein MEG3.1710010 already BLASTed\n",
      "Protein MEG4.1700110 already BLASTed\n",
      "Protein MEG4.1610110 already BLASTed\n",
      "Protein MEG15410110 already BLASTed\n",
      "Protein MEG15400100 already BLASTed\n",
      "Protein MEG3.1711000 already BLASTed\n",
      "Protein MEG4.1700100 already BLASTed\n",
      "Protein MEG8.2511000 already BLASTed\n",
      "Protein MEG151001110 already BLASTed\n",
      "Protein MEG4.1200001 already BLASTed\n",
      "Protein MEG4.1701100 already BLASTed\n",
      "Protein Smp_075800.1 already BLASTed\n",
      "Protein MEG4.11101100 already BLASTed\n",
      "Protein Smp_158420.1 already BLASTed\n",
      "Protein MEG4.11711010172 already BLASTed\n",
      "Protein MEG8.2511010 already BLASTed\n",
      "Protein MEG4.161011091 already BLASTed\n",
      "Protein MEG8.21411000 already BLASTed\n",
      "Protein MEG8.21200110 already BLASTed\n",
      "Protein MEG8.21501010 already BLASTed\n",
      "Protein MEG8.21500100 already BLASTed\n"
     ]
    }
   ],
   "source": [
    "# Ensure the output directory exists\n",
    "os.makedirs('blast_results/proteins', exist_ok=True)\n",
    "\n",
    "# find all protein sequences\n",
    "proteins = set()\n",
    "for antigen in antigens.values():\n",
    "    proteins.add((antigen.protein_id, antigen.protein_sequence))\n",
    "    \n",
    "# let's blast the protein sequences\n",
    "def blast_protein(id, seq):\n",
    "    print(f'BLASTing protein {id}')\n",
    "    result_handle = NCBIWWW.qblast(\"blastp\", \"nr\", seq)\n",
    "    with open(f'blast_results/proteins/{id}.xml', 'w') as f:\n",
    "        f.write(result_handle.read())\n",
    "    result_handle.close()\n",
    "    print(f'BLASTing protein {id} done')\n",
    "    \n",
    "\n",
    "for id, seq in proteins:\n",
    "    # check if this antigen has been blasted already\n",
    "    if os.path.exists(f'blast_results/proteins/{id}.xml'):\n",
    "        print(f'Protein {id} already BLASTed')\n",
    "        continue\n",
    "    \n",
    "    blast_protein(id, seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AlphaFold Predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
